### PR TITLE
Make `monitor` required arg of EarlyStopping callback

### PR DESCRIPTION
**Pull request by rhjohnstone**
## What does this PR do?

Makes `monitor` a required argument of `EarlyStopping` callback, as one of the deprecated behaviours mentioned in #10312.

### Does your PR introduce any breaking changes? If yes, please list them.

Yes. Users who logged the key "early_stop_on" and relied on EarlyStopping to default to monitor that key will have to change their code to `EarlyStopping(monitor="early_stop_on")`.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by awaelchli** [comment]
Great! 

There are a few tests failing because the argument is now required. These tests need to be updated. Could you do it? 
Here is the list:

```
FAILED tests/callbacks/test_early_stopping.py::test_pickling - TypeError: __i...
FAILED tests/callbacks/test_early_stopping.py::test_early_stopping_mode_options
FAILED tests/trainer/connectors/test_callback_connector.py::test_checkpoint_callbacks_are_last
FAILED tests/trainer/connectors/test_callback_connector.py::test_attach_model_callbacks
FAILED tests/trainer/connectors/test_callback_connector.py::test_attach_model_callbacks_override_info
FAILED tests/trainer/flags/test_fast_dev_run.py::test_callbacks_and_logger_not_called_with_fastdevrun[1]
FAILED tests/trainer/flags/test_fast_dev_run.py::test_callbacks_and_logger_not_called_with_fastdevrun[4]
```

If you want to check and run a test locally, you can do:

`py.test -v tests/path/to/file.py::name_of_test`

**Comment by rhjohnstone** [comment]
Thanks @awaelchli, I've fixed those tests. I now see that the docs ([this page](https://pytorch-lightning.readthedocs.io/en/stable/common/lightning_cli.html), I think) are failing because `monitor` is required. Unfortunately, I have no experience with the CLI (and hadn't ever seen those docs before!).

Is it as simple as changing
```
python ... \
    --trainer.callbacks=EarlyStopping \
    --trainer.callbacks.patience=5 \
    --trainer.callbacks=LearningRateMonitor \
    --trainer.callbacks.logging_interval=epoch
```
to
```
python ... \
    --trainer.callbacks=EarlyStopping \
    --trainer.callbacks.monitor=val_loss \
    --trainer.callbacks.patience=5 \
    --trainer.callbacks=LearningRateMonitor \
    --trainer.callbacks.logging_interval=epoch
```
etc.? I don't know if there's any special CLI behaviour when it comes to required vs optional args.

File: tests/callbacks/test_early_stopping.py [hunk 1]
```diff
@@ -34,14 +34,14 @@
 
 
 def test_early_stopping_state_key():
-    early_stopping = EarlyStopping(monitor="val_loss")
```
Comment (awaelchli) [inline tests/callbacks/test_early_stopping.py:37]: let's keep the argument named.

**Comment by awaelchli** [comment]
@rhjohnstone I updated the "breaking changes" description in the PR :)

File: docs/source/common/lightning_cli.rst [hunk 2]
```diff
@@ -625,7 +625,7 @@ This can be implemented as follows:
     class MyLightningCLI(LightningCLI):
         def add_arguments_to_parser(self, parser):
             parser.add_lightning_class_args(EarlyStopping, "my_early_stopping")
-            parser.set_defaults({"my_early_stopping.patience": 5})
+            parser.set_defaults({"monitor": "val_loss", "my_early_stopping.patience": 5})
```
Comment (awaelchli) [inline docs/source/common/lightning_cli.rst:628]: @carmocca I get 

`'No action for destination key "monitor" to set its default.'`

Can you give me a hint what is missing in the config?

Comment (carmocca) [reply docs/source/common/lightning_cli.rst:628]: Fixed

```suggestion
            parser.set_defaults({"my_early_stopping.monitor": "val_loss", "my_early_stopping.patience": 5})
```

File: tests/callbacks/test_early_stopping.py [hunk 3]
```diff
@@ -183,7 +181,7 @@ def training_epoch_end(self, outputs):
 
 
 def test_pickling(tmpdir):
-    early_stopping = EarlyStopping()
+    early_stopping = EarlyStopping("val_loss")
```
Comment (carmocca) [inline tests/callbacks/test_early_stopping.py:184]: Can we use "foo" instead of "val_loss" in all the instances where the value does not matter?

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by carmocca — APPROVED** [review]

File: tests/callbacks/test_early_stopping.py [hunk 4]
```diff
@@ -183,7 +183,7 @@ def training_epoch_end(self, outputs):
 
 
 def test_pickling(tmpdir):
-    early_stopping = EarlyStopping()
+    early_stopping = EarlyStopping("foo")
```
Comment (rohitgr7) [inline tests/callbacks/test_early_stopping.py:186]:
```suggestion
    early_stopping = EarlyStopping(monitor="foo")
```

File: tests/callbacks/test_early_stopping.py [hunk 5]
```diff
@@ -350,7 +350,7 @@ def validation_epoch_end(self, outputs):
 
 def test_early_stopping_mode_options():
     with pytest.raises(MisconfigurationException, match="`mode` can be .* got unknown_option"):
-        EarlyStopping(mode="unknown_option")
+        EarlyStopping("foo", mode="unknown_option")
```
Comment (rohitgr7) [inline tests/callbacks/test_early_stopping.py:353]:
```suggestion
        EarlyStopping(monitor="foo", mode="unknown_option")
```

File: tests/trainer/connectors/test_callback_connector.py [hunk 6]
```diff
@@ -198,8 +198,10 @@ def _attach_callbacks(trainer_callbacks, model_callbacks):
 def test_attach_model_callbacks_override_info(caplog):
     """Test that the logs contain the info about overriding callbacks returned by configure_callbacks."""
     model = LightningModule()
-    model.configure_callbacks = lambda: [LearningRateMonitor(), EarlyStopping()]
-    trainer = Trainer(enable_checkpointing=False, callbacks=[EarlyStopping(), LearningRateMonitor(), TQDMProgressBar()])
+    model.configure_callbacks = lambda: [LearningRateMonitor(), EarlyStopping("foo")]
```
Comment (rohitgr7) [inline tests/trainer/connectors/test_callback_connector.py:201]:
```suggestion
    model.configure_callbacks = lambda: [LearningRateMonitor(), EarlyStopping(monitor="foo")]
```

File: tests/trainer/connectors/test_callback_connector.py [hunk 7]
```diff
@@ -186,9 +186,9 @@ def _attach_callbacks(trainer_callbacks, model_callbacks):
         trainer_callbacks=[
             LearningRateMonitor(),
             progress_bar,
-            EarlyStopping(),
+            EarlyStopping("foo"),
             LearningRateMonitor(),
-            EarlyStopping(),
+            EarlyStopping("foo"),
```
Comment (rohitgr7) [inline tests/trainer/connectors/test_callback_connector.py:191]:
```suggestion
            EarlyStopping(monitor="foo"),
```

File: tests/trainer/connectors/test_callback_connector.py [hunk 8]
```diff
@@ -186,9 +186,9 @@ def _attach_callbacks(trainer_callbacks, model_callbacks):
         trainer_callbacks=[
             LearningRateMonitor(),
             progress_bar,
-            EarlyStopping(),
+            EarlyStopping("foo"),
```
Comment (rohitgr7) [inline tests/trainer/connectors/test_callback_connector.py:189]:
```suggestion
            EarlyStopping(monitor="foo"),
```

File: tests/trainer/connectors/test_callback_connector.py [hunk 9]
```diff
@@ -169,14 +169,14 @@ def _attach_callbacks(trainer_callbacks, model_callbacks):
...
-        trainer_callbacks=[progress_bar, EarlyStopping()],
+        trainer_callbacks=[progress_bar, EarlyStopping("foo")],
         model_callbacks=[early_stopping],
     )
     assert trainer.callbacks == [progress_bar, trainer.accumulation_scheduler, early_stopping]
 
     # multiple callbacks of the same type in trainer
     trainer = _attach_callbacks(
-        trainer_callbacks=[LearningRateMonitor(), EarlyStopping(), LearningRateMonitor(), EarlyStopping()],
+        trainer_callbacks=[LearningRateMonitor(), EarlyStopping("foo"), LearningRateMonitor(), EarlyStopping("foo")],
```
Comment (rohitgr7) [inline tests/trainer/connectors/test_callback_connector.py:179]:
```suggestion
        trainer_callbacks=[LearningRateMonitor(), EarlyStopping(monitor="foo"), LearningRateMonitor(), EarlyStopping(monitor="foo")],
```

File: tests/trainer/connectors/test_callback_connector.py [hunk 10]
```diff
@@ -169,14 +169,14 @@ def _attach_callbacks(trainer_callbacks, model_callbacks):
 
     # same callback type twice, different instance
     trainer = _attach_callbacks(
-        trainer_callbacks=[progress_bar, EarlyStopping()],
+        trainer_callbacks=[progress_bar, EarlyStopping("foo")],
```
Comment (rohitgr7) [inline tests/trainer/connectors/test_callback_connector.py:172]:
```suggestion
        trainer_callbacks=[progress_bar, EarlyStopping(monitor="foo")],
```

File: tests/trainer/connectors/test_callback_connector.py [hunk 11]
```diff
@@ -154,7 +154,7 @@ def _attach_callbacks(trainer_callbacks, model_callbacks):
         cb_connector._attach_model_callbacks()
         return trainer
 
-    early_stopping = EarlyStopping()
+    early_stopping = EarlyStopping("foo")
```
Comment (rohitgr7) [inline tests/trainer/connectors/test_callback_connector.py:157]:
```suggestion
    early_stopping = EarlyStopping(monitor="foo")
```

File: tests/trainer/connectors/test_callback_connector.py [hunk 12]
```diff
@@ -33,7 +33,7 @@ def test_checkpoint_callbacks_are_last(tmpdir):
     checkpoint1 = ModelCheckpoint(tmpdir)
     checkpoint2 = ModelCheckpoint(tmpdir)
     model_summary = ModelSummary()
-    early_stopping = EarlyStopping()
+    early_stopping = EarlyStopping("foo")
```
Comment (rohitgr7) [inline tests/trainer/connectors/test_callback_connector.py:36]:
```suggestion
    early_stopping = EarlyStopping(monitor="foo")
```

**Review by rohitgr7 — APPROVED** [review]
small update to use it as kwarg

File: tests/trainer/connectors/test_callback_connector.py [hunk 13]
```diff
@@ -198,8 +198,10 @@ def _attach_callbacks(trainer_callbacks, model_callbacks):
 def test_attach_model_callbacks_override_info(caplog):
     """Test that the logs contain the info about overriding callbacks returned by configure_callbacks."""
     model = LightningModule()
-    model.configure_callbacks = lambda: [LearningRateMonitor(), EarlyStopping()]
-    trainer = Trainer(enable_checkpointing=False, callbacks=[EarlyStopping(), LearningRateMonitor(), TQDMProgressBar()])
+    model.configure_callbacks = lambda: [LearningRateMonitor(), EarlyStopping(monitor="foo")]
+    trainer = Trainer(
+        enable_checkpointing=False, callbacks=[EarlyStopping("foo"), LearningRateMonitor(), TQDMProgressBar()]
```
Comment (rohitgr7) [inline tests/trainer/connectors/test_callback_connector.py:203]:
```suggestion
        enable_checkpointing=False, callbacks=[EarlyStopping(monitor="foo"), LearningRateMonitor(), TQDMProgressBar()]
```

**Pull request merged**